src/ship/PlayerShipController.cpp

```diff
@@ -9,13 +9,138 @@
 #include "Input.h"
 #include "Pi.h"
 #include "Player.h"
+#include "SDL_keycode.h"
 #include "Ship.h"
 #include "Space.h"
+#include "SystemView.h"
 #include "WorldView.h"
 #include "core/OS.h"
 #include "lua/LuaObject.h"
+#include "ship/ShipController.h"
 
 #include <algorithm>
+#include <optional>
+
+// so that we can summarize the controller action from all subsystems
+// vector3d (0.0) means we want zero speed, for example, so I made optional
+// so that the system can say that it does not want to touch this parameter at all
+struct PlayerShipController::TotalDesiredAction {
+	std::optional<vector3d> velLinear = {};	 // in ship's frame
+	std::optional<vector3d> velAngular = {}; // in ship's local space
+	std::optional<vector3d> thrLinear = {};
+	std::optional<vector3d> thrAngular = {};
+	vector3d angPower = vector3d(1.0);
+	vector3d linPower = vector3d(1.0);
+};
```

Comment:
Instead of using `std::optional` here which wastes an extra 7 bytes for each vector3d member, this struct can be collapsed into four vectors and two bools. std::optional is a good fit for function parameters and return types where you need to bundle both value and state into a single return type, but for general-purpose data members it is less useful; in this case I think it's probably a better fit to think of the velocity-vs-thrust distinction as two exclusive modes rather than two optional values.

```suggestion
struct PlayerShipController::TotalDesiredAction {
	vector3d desiredLinear = {};	 // in ship's frame
	vector3d desiredAngular = {}; // in ship's local space
	vector3d angPower = vector3d(1.0);
	vector3d linPower = vector3d(1.0);
	bool desireLinVel = false;
	bool desireAngVel = false;
};
```